### PR TITLE
Check uniqueness of impl items by trait item when applicable.

### LINKED ISSUE #71614: Declarative macros: no error about duplicate trait items
I tried this code on the latest nightly:

```rust
#![feature(decl_macro)]

trait Trait {
    type T;
}

macro trait_impl {
    () => {
        type T = ();
    }
}

impl Trait for i32 {
    trait_impl!();
    type T = ();
}

```

I expected to see this happen: an error about `type T` being defined twice in the same impl. (This is what happens with `macro_rules! trait_impl`.)

Instead, this happened: the code just compiles. The same also works when considering a function rather than a type.

### PR DESCRIPTION
**Pull request by cjgillot**
When checking uniqueness of item names in impl blocks, we currently use the same definition of hygiene as for toplevel items.  This means that a plain item and one generated by a macro 2.0 do not collide.

This hygiene rule does not match with how impl items resolve to associated trait items. As a consequence, we misdiagnose the trait impls.

This PR proposes to consider that trait impl items are uses of the corresponding trait items during resolution, instead of checking for duplicates later. An error is emitted when a trait impl item is used twice.

There should be no stable breakage, since macros 2.0 are still unstable.

r? @petrochenkov 
cc @RalfJung 

Fixes https://github.com/rust-lang/rust/issues/71614.

### CODE REVIEW

File: compiler/rustc_typeck/src/impl_wf_check.rs [hunk 1]
```diff
@@ -207,7 +207,15 @@ fn enforce_impl_items_are_distinct(tcx: TyCtxt<'_>, impl_def_id: LocalDefId) {
         };
         let span = tcx.def_span(impl_item_ref);
         let ident = impl_item.ident(tcx);
-        match seen_items.entry(ident.normalize_to_macros_2_0()) {
+
+        // For trait impls, we check that there is a single implementation for each trait item.
+        // For inherent impls, we check that there is no two items with the same name.
```
Comment (RalfJung) [inline compiler/rustc_typeck/src/impl_wf_check.rs:212]: Might be worth saying explicitly that "name" here includes the hygiene disambiguator or whatever it is called.

File: src/test/ui/hygiene/impl_items-2.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+struct Type;
+
+// Check that we do not error with inherent impls.
+impl Type {
+    trait_impl!();
+    fn foo() {}
+}
+
+fn main() {}
```
Comment (RalfJung) [inline src/test/ui/hygiene/impl_items-2.rs:26]: Would be good to have a test ensuring that if `main` calls `Type::foo`, it is the one from *outside* the macro. Or are there such tests already?

File: compiler/rustc_typeck/src/impl_wf_check.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_typeck/src/impl_wf_check.rs:212]:
```suggestion
        // For inherent impls, we check that there is no two items with the same name, including hygiene.
```

File: src/test/ui/hygiene/impl_items-2.rs (hunk 2 above)
Comment (cjgillot) [reply src/test/ui/hygiene/impl_items-2.rs:26]: There is already `src/test/ui/hygiene/impl_items.rs` which tests according to types.

**Comment by petrochenkov** [comment]
I need to refresh hygiene details in my memory before reviewing this PR and https://github.com/rust-lang/rust/pull/99445, I'll try to do it some time during the next couple of weeks.

**Comment by petrochenkov** [comment]
Let's focus on trait impls first (as opposed to inherent impls).

Names in a trait impl (trait impl items) are considered *uses* referring to their *definitions* in the trait.
So they are resolved as uses - code in `fn check_trait_item` does it manually, but it should be equivalent to `resolve_ident_in_module` (with the trait being the module).

This PR doesn't change that scheme.
So I guess for traits we can check both completeness of trait items (that should consider specialization?) and the lack of duplicates in the same place, using only their `DefId`s obtained early during name resolution and not names.

**Comment by cjgillot** [comment]
@petrochenkov I split the PR in 2 parts.
This PR keeps the change for trait impl items.
https://github.com/rust-lang/rust/pull/102856 moves the check for duplicate inherent impl items work with overlap check.

File: compiler/rustc_hir_analysis/src/impl_wf_check.rs [hunk 3]
```diff
@@ -207,7 +210,9 @@ fn enforce_impl_items_are_distinct(tcx: TyCtxt<'_>, impl_def_id: LocalDefId) {
         };
         let span = tcx.def_span(impl_item_ref);
         let ident = impl_item.ident(tcx);
-        match seen_items.entry(ident.normalize_to_macros_2_0()) {
+
+        let norm_ident = ident.normalize_to_macros_2_0();
+        match seen_items.entry(norm_ident) {
```
Comment (petrochenkov) [inline compiler/rustc_hir_analysis/src/impl_wf_check.rs:215]: Nit: unnecessary change.

File: compiler/rustc_resolve/src/late.rs [hunk 4]
```diff
@@ -2768,7 +2777,25 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
         };
 
         let res = binding.res();
-        let Res::Def(def_kind, _) = res else { bug!() };
+        let Res::Def(def_kind, id_in_trait) = res else { bug!() };
+
+        match seen_trait_items[ns].entry(id_in_trait) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late.rs:2782]:
```suggestion
        match seen_trait_items.entry(id_in_trait) {
```
I don't think `PerNS` is necessary here?
`DefId`s of the trait items should be disjoint anyway.

**Comment by petrochenkov** [comment]
r=me after squashing commits.
@rustbot author
